tests/raftstore/transport_simulate.rs

```diff
@@ -459,3 +459,68 @@ impl Filter<StoreMsg> for FilterLeadingDuplicatedSnap {
         if res.is_err() && dropped { Ok(()) } else { res }
     }
 }
+
+/// `RandomLatencyFilter` is a transport filter to simulate randomized network latency.
+/// Based on a randomized rate, `RandomLatencyFilter` will decide whether to delay
+/// the sending of any message. It's could be used to simulate the message sending
+/// in a network with random latency, where messages could be delayed, disordered or lost.
+pub struct RandomLatencyFilter {
+    delay_rate: u32,
+    delayed_msgs: Mutex<Vec<RaftMessage>>,
+}
+
+impl RandomLatencyFilter {
+    pub fn new(rate: u32) -> RandomLatencyFilter {
+        RandomLatencyFilter {
+            delay_rate: rate,
+            delayed_msgs: Mutex::new(vec![]),
+        }
+    }
+
+    fn will_delay(&self, _: &RaftMessage) -> bool {
+        rand::random::<u32>() % 100u32 >= self.delay_rate
+    }
+}
+
+impl Filter<RaftMessage> for RandomLatencyFilter {
+    fn before(&self, msgs: &mut Vec<RaftMessage>) {
```

Comment:
If msgs has only one message, we will return error for this send, and raft will re-send the message again. So this may be not for `delay`, but `duplicated`.